// openvidu-angular/esm2020/lib/services/action/action.service.mjs
import { Injectable } from '@angular/core';
import { DeleteDialogComponent } from '../../components/dialogs/delete-recording.component';
import { DialogTemplateComponent } from '../../components/dialogs/dialog.component';
import { ProFeatureDialogTemplateComponent } from '../../components/dialogs/pro-feature-dialog.component';
import { RecordingDialogComponent } from '../../components/dialogs/recording-dialog.component';
import * as i0 from "@angular/core";
import * as i1 from "@angular/material/snack-bar";
import * as i2 from "@angular/material/dialog";
/**
 * @internal
 */
export class ActionService {
    constructor(snackBar, dialog) {
        this.snackBar = snackBar;
        this.dialog = dialog;
    }
    launchNotification(options, callback) {
        if (!options.config) {
            options.config = {
                duration: 3000,
                verticalPosition: 'top',
                horizontalPosition: 'end'
            };
        }
        const notification = this.snackBar.open(options.message, options.buttonActionText, options.config);
        if (callback) {
            notification.onAction().subscribe(() => {
                callback();
            });
        }
    }
    openDialog(titleMessage, descriptionMessage, allowClose = true) {
        try {
            this.closeDialog();
        }
        catch (error) {
        }
        finally {
            const config = {
                minWidth: '250px',
                data: { title: titleMessage, description: descriptionMessage, showActionButtons: allowClose },
                disableClose: !allowClose
            };
            this.dialogRef = this.dialog.open(DialogTemplateComponent, config);
            this.dialogSubscription = this.dialogRef.afterClosed().subscribe((result) => {
                this.dialogRef = undefined;
            });
        }
    }
    openProFeatureDialog(titleMessage, descriptionMessage, allowClose = true) {
        try {
            this.closeDialog();
        }
        catch (error) {
        }
        finally {
            const config = {
                minWidth: '250px',
                data: { title: titleMessage, description: descriptionMessage, showActionButtons: allowClose },
                disableClose: !allowClose
            };
            this.dialogRef = this.dialog.open(ProFeatureDialogTemplateComponent, config);
            this.dialogSubscription = this.dialogRef.afterClosed().subscribe((result) => {
                this.dialogRef = undefined;
            });
        }
    }
    openDeleteRecordingDialog(succsessCallback) {
        try {
            this.closeDialog();
        }
        catch (error) {
        }
        finally {
            this.dialogRef = this.dialog.open(DeleteDialogComponent);
            this.dialogSubscription = this.dialogRef.afterClosed().subscribe((result) => {
                if (result) {
                    succsessCallback();
                }
            });
        }
    }
    openRecordingPlayerDialog(src, allowClose = true) {
        try {
            this.closeDialog();
        }
        catch (error) {
        }
        finally {
            const config = {
                minWidth: '250px',
                data: { src, showActionButtons: allowClose },
                disableClose: !allowClose
            };
            this.dialogRef = this.dialog.open(RecordingDialogComponent, config);
        }
    }
    closeDialog() {
        this.dialogRef?.close();
        if (this.dialogSubscription)
            this.dialogSubscription.unsubscribe();
    }
}
ActionService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.8", ngImport: i0, type: ActionService, deps: [{ token: i1.MatSnackBar }, { token: i2.MatDialog }], target: i0.ɵɵFactoryTarget.Injectable });
ActionService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "14.2.8", ngImport: i0, type: ActionService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.8", ngImport: i0, type: ActionService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: function () { return [{ type: i1.MatSnackBar }, { type: i2.MatDialog }]; } });
