// openvidu-angular/esm2020/lib/components/video/video.component.mjs
import { Component, Input, ViewChild } from '@angular/core';
import { VideoType } from '../../models/video-type.model';
import * as i0 from "@angular/core";
/**
 * @internal
 */
export class VideoComponent {
    constructor() {
        this.type = VideoType.CAMERA;
    }
    ngAfterViewInit() {
        setTimeout(() => {
            if (this._streamManager && this._videoElement) {
                this.updateVideoStyles();
                this._streamManager.addVideoElement(this._videoElement.nativeElement);
            }
        });
    }
    set videoElement(element) {
        this._videoElement = element;
    }
    set streamManager(streamManager) {
        if (streamManager) {
            this._streamManager = streamManager;
            if (!!this._videoElement && this._streamManager) {
                this.updateVideoStyles();
                this._streamManager.addVideoElement(this._videoElement.nativeElement);
            }
        }
    }
    updateVideoStyles() {
        this.type = this._streamManager?.stream?.typeOfVideo;
        if (this.type === VideoType.SCREEN) {
            this._videoElement.nativeElement.style.objectFit = 'contain';
            this._videoElement.nativeElement.classList.add('screen-type');
        }
        else {
            this._videoElement.nativeElement.style.objectFit = 'cover';
            this._videoElement.nativeElement.classList.add('camera-type');
        }
    }
}
VideoComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.8", ngImport: i0, type: VideoComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
VideoComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "14.2.8", type: VideoComponent, selector: "ov-video", inputs: { mutedSound: "mutedSound", streamManager: "streamManager" }, viewQueries: [{ propertyName: "videoElement", first: true, predicate: ["videoElement"], descendants: true }], ngImport: i0, template: `
		<video
			class="OT_video-element"
			#videoElement
			[attr.id]="streamManager && _streamManager.stream ? 'video-' + _streamManager.stream.streamId : 'video-undefined'"
			[muted]="mutedSound"
		></video>
	`, isInline: true, styles: ["video{object-fit:cover;width:100%;height:100%;color:#fff;margin:0;padding:0;border:0;font-size:100%;border-radius:var(--ov-video-radius);background-color:#000}\n"] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.8", ngImport: i0, type: VideoComponent, decorators: [{
            type: Component,
            args: [{ selector: 'ov-video', template: `
		<video
			class="OT_video-element"
			#videoElement
			[attr.id]="streamManager && _streamManager.stream ? 'video-' + _streamManager.stream.streamId : 'video-undefined'"
			[muted]="mutedSound"
		></video>
	`, styles: ["video{object-fit:cover;width:100%;height:100%;color:#fff;margin:0;padding:0;border:0;font-size:100%;border-radius:var(--ov-video-radius);background-color:#000}\n"] }]
        }], propDecorators: { mutedSound: [{
                type: Input
            }], videoElement: [{
                type: ViewChild,
                args: ['videoElement', { static: false }]
            }], streamManager: [{
                type: Input
            }] } });
