// openvidu-angular/esm2020/lib/pipes/participant.pipe.mjs
import { Pipe } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "../services/translate/translate.service";
export class ParticipantStreamsPipe {
    constructor() { }
    transform(participants) {
        let streams = [];
        if (participants && Object.keys(participants).length > 0) {
            if (Array.isArray(participants)) {
                participants.forEach((p) => {
                    streams = streams.concat(p.getAvailableConnections());
                });
            }
            else {
                streams = participants.getAvailableConnections();
            }
        }
        return streams;
    }
}
ParticipantStreamsPipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.8", ngImport: i0, type: ParticipantStreamsPipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe });
ParticipantStreamsPipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "14.2.8", ngImport: i0, type: ParticipantStreamsPipe, name: "streams" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.8", ngImport: i0, type: ParticipantStreamsPipe, decorators: [{
            type: Pipe,
            args: [{ name: 'streams' }]
        }], ctorParameters: function () { return []; } });
/**
 * @internal
 */
export class StreamTypesEnabledPipe {
    constructor(translateService) {
        this.translateService = translateService;
    }
    transform(participant) {
        let result = '';
        let activeStreams = participant?.getConnectionTypesActive().toString();
        const activeStreamsArr = activeStreams.split(',');
        activeStreamsArr.forEach((type, index) => {
            result += this.translateService.translate(`PANEL.PARTICIPANTS.${type}`);
            if (activeStreamsArr.length > 0 && index < activeStreamsArr.length - 1) {
                result += ', ';
            }
        });
        return `(${result})`;
    }
}
StreamTypesEnabledPipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.8", ngImport: i0, type: StreamTypesEnabledPipe, deps: [{ token: i1.TranslateService }], target: i0.ɵɵFactoryTarget.Pipe });
StreamTypesEnabledPipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "14.2.8", ngImport: i0, type: StreamTypesEnabledPipe, name: "streamTypesEnabled" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.8", ngImport: i0, type: StreamTypesEnabledPipe, decorators: [{
            type: Pipe,
            args: [{ name: 'streamTypesEnabled' }]
        }], ctorParameters: function () { return [{ type: i1.TranslateService }]; } });
