// openvidu-angular/esm2020/lib/pipes/recording.pipe.mjs
import { Pipe } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * @internal
 */
export class DurationFromSecondsPipe {
    transform(durationInSeconds) {
        if (durationInSeconds < 60) {
            return Math.floor(durationInSeconds) + 's';
        }
        else if (durationInSeconds < 3600) {
            return Math.floor(durationInSeconds / 60) + 'm ' + Math.floor(durationInSeconds % 60) + 's';
        }
        else {
            const hours = Math.floor(durationInSeconds / 3600);
            const minutes = Math.floor((durationInSeconds - hours * 3600) / 60);
            return hours + 'h ' + minutes + 'm';
        }
    }
}
DurationFromSecondsPipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.8", ngImport: i0, type: DurationFromSecondsPipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe });
DurationFromSecondsPipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "14.2.8", ngImport: i0, type: DurationFromSecondsPipe, name: "duration" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.8", ngImport: i0, type: DurationFromSecondsPipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'duration'
                }]
        }] });
/**
 * @internal
 */
export class SearchByStringPropertyPipe {
    transform(items, props) {
        if (!items || !props || props.properties.length === 0 || !props.filter) {
            return items;
        }
        return items.filter(item => {
            return props.properties.some(prop => {
                const multipleProps = prop.split('.');
                let recursiveProp = item;
                try {
                    multipleProps.forEach(p => {
                        recursiveProp = recursiveProp[p];
                        if (recursiveProp === null || recursiveProp === undefined) {
                            throw new Error('Property not found');
                        }
                    });
                }
                catch (error) {
                    return false;
                }
                return recursiveProp.indexOf(props.filter) !== -1;
            });
        });
    }
}
SearchByStringPropertyPipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.8", ngImport: i0, type: SearchByStringPropertyPipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe });
SearchByStringPropertyPipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "14.2.8", ngImport: i0, type: SearchByStringPropertyPipe, name: "searchByStringProperty" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.8", ngImport: i0, type: SearchByStringPropertyPipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'searchByStringProperty'
                }]
        }] });
/**
 * @internal
 */
export class ThumbnailFromUrlPipe {
    transform(url) {
        if (url.includes('.mp4')) {
            const lastPart = url.split('/').pop();
            let thumbnailUrl = lastPart?.replace('mp4', 'jpg');
            thumbnailUrl = `recordings/${thumbnailUrl?.split('.')[0]}/${thumbnailUrl}`;
            return thumbnailUrl;
        }
        return url;
    }
}
ThumbnailFromUrlPipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.8", ngImport: i0, type: ThumbnailFromUrlPipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe });
ThumbnailFromUrlPipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "14.2.8", ngImport: i0, type: ThumbnailFromUrlPipe, name: "thumbnailUrl" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.8", ngImport: i0, type: ThumbnailFromUrlPipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'thumbnailUrl'
                }]
        }] });
