// openvidu-angular/esm2020/lib/services/cdk-overlay/cdk-overlay.service.mjs
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "../../config/custom-cdk-overlay";
/**
 * @internal
 */
export class CdkOverlayService {
    constructor(cdkOverlayModel) {
        this.cdkOverlayModel = cdkOverlayModel;
    }
    setSelector(selector) {
        this.cdkOverlayModel.setSelector(selector);
    }
}
CdkOverlayService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.8", ngImport: i0, type: CdkOverlayService, deps: [{ token: i1.CdkOverlayContainer }], target: i0.ɵɵFactoryTarget.Injectable });
CdkOverlayService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "14.2.8", ngImport: i0, type: CdkOverlayService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.8", ngImport: i0, type: CdkOverlayService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: function () { return [{ type: i1.CdkOverlayContainer }]; } });
