// openvidu-angular/esm2020/lib/config/custom-flexlayout-breakpoints.mjs
import { BREAKPOINT, LayoutDirective } from '@angular/flex-layout';
import { Directive } from '@angular/core';
import * as i0 from "@angular/core";
const LANDSCAPE_BREAKPOINTS = [
    {
        alias: 'landscape',
        suffix: 'Landscape',
        mediaQuery: 'screen and (orientation: landscape)',
        overlapping: false,
        priority: 2001
    }
];
export const CustomBreakPointsProvider = {
    provide: BREAKPOINT,
    useValue: LANDSCAPE_BREAKPOINTS,
    multi: true
};
const selector = `[fxLayout.landscape]`;
const inputs = ['fxLayout.landscape'];
export class CustomLayoutExtensionDirective extends LayoutDirective {
    constructor() {
        super(...arguments);
        this.inputs = inputs;
    }
}
CustomLayoutExtensionDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.8", ngImport: i0, type: CustomLayoutExtensionDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive });
CustomLayoutExtensionDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "14.2.8", type: CustomLayoutExtensionDirective, selector: "[fxLayout.landscape]", inputs: { "fxLayout.landscape": "fxLayout.landscape" }, usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.8", ngImport: i0, type: CustomLayoutExtensionDirective, decorators: [{
            type: Directive,
            args: [{ selector, inputs }]
        }] });
