// openvidu-angular/esm2020/lib/config/custom-cdk-overlay.mjs
import { OverlayContainer } from '@angular/cdk/overlay';
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class CdkOverlayContainer extends OverlayContainer {
    constructor() {
        super(...arguments);
        this.containerSelector = '.sidenav-main';
        this.customClass = 'cdk-overlay-container';
    }
    _createContainer() {
        const container = document.createElement('div');
        container.classList.add(this.customClass);
        let element = this.getElement(this.containerSelector);
        if (!element) {
            element = this.getElement('body');
        }
        this._containerElement = element.appendChild(container);
    }
    setSelector(selector) {
        const overlayElement = this.getElement('.' + this.customClass);
        if (overlayElement && this.containerSelector !== selector) {
            const newContainerOverlayContainer = this.getElement(selector);
            this.containerSelector = selector;
            newContainerOverlayContainer?.appendChild(overlayElement);
        }
    }
    getElement(selector) {
        return document.querySelector(selector);
    }
}
CdkOverlayContainer.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.8", ngImport: i0, type: CdkOverlayContainer, deps: null, target: i0.ɵɵFactoryTarget.Injectable });
CdkOverlayContainer.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "14.2.8", ngImport: i0, type: CdkOverlayContainer });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.8", ngImport: i0, type: CdkOverlayContainer, decorators: [{
            type: Injectable
        }] });
