// openvidu-angular/esm2020/lib/components/settings/lang-selector/lang-selector.component.mjs
import { Component, Output, ViewChild, EventEmitter } from '@angular/core';
import { MatMenuTrigger } from '@angular/material/menu';
import { MatSelect } from '@angular/material/select';
import * as i0 from "@angular/core";
import * as i1 from "../../../services/translate/translate.service";
import * as i2 from "../../../services/storage/storage.service";
import * as i3 from "@angular/common";
import * as i4 from "@angular/material/button";
import * as i5 from "@angular/material/icon";
import * as i6 from "@angular/material/menu";
/**
 * @internal
 */
export class LangSelectorComponent {
    constructor(translateService, storageSrv) {
        this.translateService = translateService;
        this.storageSrv = storageSrv;
        this.onLangSelectorClicked = new EventEmitter();
        this.languages = [];
    }
    ngOnInit() {
        this.languages = this.translateService.getLanguagesInfo();
        this.langSelected = this.translateService.getLangSelected();
    }
    ngAfterViewInit() {
        this.menuTrigger?.menuOpened.subscribe(() => {
            this.onLangSelectorClicked.emit();
        });
        this.matSelect?.openedChange.subscribe(() => {
            this.onLangSelectorClicked.emit();
        });
    }
    onLangSelected(lang) {
        this.translateService.setLanguage(lang);
        this.storageSrv.setLang(lang);
        this.langSelected = this.translateService.getLangSelected();
    }
}
LangSelectorComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.8", ngImport: i0, type: LangSelectorComponent, deps: [{ token: i1.TranslateService }, { token: i2.StorageService }], target: i0.ɵɵFactoryTarget.Component });
LangSelectorComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "14.2.8", type: LangSelectorComponent, selector: "ov-lang-selector", outputs: { onLangSelectorClicked: "onLangSelectorClicked" }, viewQueries: [{ propertyName: "menuTrigger", first: true, predicate: MatMenuTrigger, descendants: true }, { propertyName: "matSelect", first: true, predicate: MatSelect, descendants: true }], ngImport: i0, template: "<button mat-flat-button [matMenuTriggerFor]=\"menu\" class=\"lang-button\">\n\t<span>{{langSelected?.name}}</span>\n\t<mat-icon>expand_more</mat-icon>\n</button>\n<mat-menu #menu=\"matMenu\">\n\t<button mat-menu-item *ngFor=\"let lang of languages\" (click)=\"onLangSelected(lang.ISO)\">\n\t\t<span>{{lang.name}}</span>\n\t</button>\n</mat-menu>", styles: [".lang-button{background-color:var(--ov-logo-background-color);color:var(--ov-text-color)}\n"], dependencies: [{ kind: "directive", type: i3.NgForOf, selector: "[ngFor][ngForOf]", inputs: ["ngForOf", "ngForTrackBy", "ngForTemplate"] }, { kind: "component", type: i4.MatButton, selector: "button[mat-button], button[mat-raised-button], button[mat-icon-button],             button[mat-fab], button[mat-mini-fab], button[mat-stroked-button],             button[mat-flat-button]", inputs: ["disabled", "disableRipple", "color"], exportAs: ["matButton"] }, { kind: "component", type: i5.MatIcon, selector: "mat-icon", inputs: ["color", "inline", "svgIcon", "fontSet", "fontIcon"], exportAs: ["matIcon"] }, { kind: "component", type: i6.MatMenu, selector: "mat-menu", exportAs: ["matMenu"] }, { kind: "component", type: i6.MatMenuItem, selector: "[mat-menu-item]", inputs: ["disabled", "disableRipple", "role"], exportAs: ["matMenuItem"] }, { kind: "directive", type: i6.MatMenuTrigger, selector: "[mat-menu-trigger-for], [matMenuTriggerFor]", exportAs: ["matMenuTrigger"] }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.8", ngImport: i0, type: LangSelectorComponent, decorators: [{
            type: Component,
            args: [{ selector: 'ov-lang-selector', template: "<button mat-flat-button [matMenuTriggerFor]=\"menu\" class=\"lang-button\">\n\t<span>{{langSelected?.name}}</span>\n\t<mat-icon>expand_more</mat-icon>\n</button>\n<mat-menu #menu=\"matMenu\">\n\t<button mat-menu-item *ngFor=\"let lang of languages\" (click)=\"onLangSelected(lang.ISO)\">\n\t\t<span>{{lang.name}}</span>\n\t</button>\n</mat-menu>", styles: [".lang-button{background-color:var(--ov-logo-background-color);color:var(--ov-text-color)}\n"] }]
        }], ctorParameters: function () { return [{ type: i1.TranslateService }, { type: i2.StorageService }]; }, propDecorators: { onLangSelectorClicked: [{
                type: Output
            }], menuTrigger: [{
                type: ViewChild,
                args: [MatMenuTrigger]
            }], matSelect: [{
                type: ViewChild,
                args: [MatSelect]
            }] } });
