// openvidu-angular/esm2020/lib/services/platform/platform.service.mjs
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * @internal
 */
export class PlatformService {
    constructor() { }
    isMobile() {
        return this.isAndroid() || this.isIos();
    }
    isFirefox() {
        return /Firefox[\/\s](\d+\.\d+)/.test(navigator.userAgent);
    }
    isAndroid() {
        return /\b(\w*Android\w*)\b/.test(navigator.userAgent) && /\b(\w*Mobile\w*)\b/.test(navigator.userAgent);
    }
    isIos() {
        return this.isIPhoneOrIPad(navigator?.userAgent);
    }
    isIPhoneOrIPad(userAgent) {
        const isIPad = /\b(\w*Macintosh\w*)\b/.test(userAgent);
        const isIPhone = /\b(\w*iPhone\w*)\b/.test(userAgent) && /\b(\w*Mobile\w*)\b/.test(userAgent);
        // && /\b(\w*iPhone\w*)\b/.test(navigator.platform);
        const isTouchable = 'ontouchend' in document;
        return (isIPad || isIPhone) && isTouchable;
    }
    isSafariIos() {
        return this.isIos() && this.isIOSWithSafari(navigator?.userAgent);
    }
    isIOSWithSafari(userAgent) {
        return (/\b(\w*Apple\w*)\b/.test(navigator.vendor) &&
            /\b(\w*Safari\w*)\b/.test(userAgent) &&
            !/\b(\w*CriOS\w*)\b/.test(userAgent) &&
            !/\b(\w*FxiOS\w*)\b/.test(userAgent));
    }
}
PlatformService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.8", ngImport: i0, type: PlatformService, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
PlatformService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "14.2.8", ngImport: i0, type: PlatformService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.8", ngImport: i0, type: PlatformService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: function () { return []; } });
