// openvidu-angular/esm2020/lib/services/config/openvidu-angular.config.service.mjs
import { Inject, Injectable } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import * as i0 from "@angular/core";
// import { version } from '../../../../package.json';
/**
 * @internal
 */
export class OpenViduAngularConfigService {
    constructor(config) {
        this.minimal = new BehaviorSubject(false);
        this.participantName = new BehaviorSubject('');
        this.prejoin = new BehaviorSubject(true);
        this.videoMuted = new BehaviorSubject(false);
        this.audioMuted = new BehaviorSubject(false);
        this.screenshareButton = new BehaviorSubject(true);
        this.fullscreenButton = new BehaviorSubject(true);
        this.captionsButton = new BehaviorSubject(true);
        this.toolbarSettingsButton = new BehaviorSubject(true);
        this.leaveButton = new BehaviorSubject(true);
        this.participantsPanelButton = new BehaviorSubject(true);
        this.chatPanelButton = new BehaviorSubject(true);
        this.activitiesPanelButton = new BehaviorSubject(true);
        this.displaySessionName = new BehaviorSubject(true);
        this.displayLogo = new BehaviorSubject(true);
        this.displayParticipantName = new BehaviorSubject(true);
        this.displayAudioDetection = new BehaviorSubject(true);
        this.streamSettingsButton = new BehaviorSubject(true);
        this.participantItemMuteButton = new BehaviorSubject(true);
        this.backgroundEffectsButton = new BehaviorSubject(true);
        this.recordingsList = new BehaviorSubject([]);
        this.recordingButton = new BehaviorSubject(true);
        this.recordingActivity = new BehaviorSubject(true);
        this.recordingError = new BehaviorSubject(null);
        this.adminRecordingsList = new BehaviorSubject([]);
        this.adminLoginError = new BehaviorSubject(null);
        this.configuration = config;
        console.log(this.configuration);
        if (this.isProduction())
            console.log('OpenVidu Angular Production Mode');
        this.minimalObs = this.minimal.asObservable();
        this.participantNameObs = this.participantName.asObservable();
        this.prejoinObs = this.prejoin.asObservable();
        this.videoMutedObs = this.videoMuted.asObservable();
        this.audioMutedObs = this.audioMuted.asObservable();
        //Toolbar observables
        this.screenshareButtonObs = this.screenshareButton.asObservable();
        this.fullscreenButtonObs = this.fullscreenButton.asObservable();
        this.backgroundEffectsButtonObs = this.backgroundEffectsButton.asObservable();
        this.leaveButtonObs = this.leaveButton.asObservable();
        this.participantsPanelButtonObs = this.participantsPanelButton.asObservable();
        this.chatPanelButtonObs = this.chatPanelButton.asObservable();
        this.activitiesPanelButtonObs = this.activitiesPanelButton.asObservable();
        this.displaySessionNameObs = this.displaySessionName.asObservable();
        this.displayLogoObs = this.displayLogo.asObservable();
        this.recordingButtonObs = this.recordingButton.asObservable();
        this.toolbarSettingsButtonObs = this.toolbarSettingsButton.asObservable();
        this.captionsButtonObs = this.captionsButton.asObservable();
        //Stream observables
        this.displayParticipantNameObs = this.displayParticipantName.asObservable();
        this.displayAudioDetectionObs = this.displayAudioDetection.asObservable();
        this.streamSettingsButtonObs = this.streamSettingsButton.asObservable();
        // Participant item observables
        this.participantItemMuteButtonObs = this.participantItemMuteButton.asObservable();
        // Recording activity observables
        this.recordingActivityObs = this.recordingActivity.asObservable();
        this.recordingsListObs = this.recordingsList.asObservable();
        this.recordingErrorObs = this.recordingError.asObservable();
        // Admin dashboard
        this.adminRecordingsListObs = this.adminRecordingsList.asObservable();
        this.adminLoginErrorObs = this.adminLoginError.asObservable();
    }
    getConfig() {
        return this.configuration;
    }
    isProduction() {
        return this.configuration?.production || false;
    }
    hasParticipantFactory() {
        return typeof this.getConfig().participantFactory === 'function';
    }
    getParticipantFactory() {
        return this.getConfig().participantFactory;
    }
}
OpenViduAngularConfigService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.8", ngImport: i0, type: OpenViduAngularConfigService, deps: [{ token: 'OPENVIDU_ANGULAR_CONFIG' }], target: i0.ɵɵFactoryTarget.Injectable });
OpenViduAngularConfigService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "14.2.8", ngImport: i0, type: OpenViduAngularConfigService });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.8", ngImport: i0, type: OpenViduAngularConfigService, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Inject,
                    args: ['OPENVIDU_ANGULAR_CONFIG']
                }] }]; } });
