// openvidu-angular/esm2020/lib/components/panel/participants-panel/participant-panel-item/participant-panel-item.component.mjs
import { ChangeDetectionStrategy, Component, ContentChild, Input, TemplateRef } from '@angular/core';
import { ParticipantPanelItemElementsDirective } from '../../../../directives/template/openvidu-angular.directive';
import * as i0 from "@angular/core";
import * as i1 from "../../../../services/config/openvidu-angular.config.service";
import * as i2 from "../../../../services/participant/participant.service";
import * as i3 from "@angular/common";
import * as i4 from "@angular/material/button";
import * as i5 from "@angular/material/icon";
import * as i6 from "@angular/material/core";
import * as i7 from "@angular/material/list";
import * as i8 from "../../../../pipes/participant.pipe";
/**
 *
 * The **ParticipantPanelItemComponent** is hosted inside of the {@link ParticipantsPanelComponent}.
 * It is in charge of displaying the participants information inside of the ParticipansPanelComponent.
 *
 * <div class="custom-table-container">
 * <div>
 *  <h3>API Directives</h3>
 *
 * This component allows us to show or hide certain HTML elements with the following {@link https://angular.io/guide/attribute-directives Angular attribute directives}
 * with the aim of fully customizing the ToolbarComponent.
 *
 * | **Name**                  | **Type**  | **Reference**                                   |
 * | :----------------------------: | :-------: | :---------------------------------------------: |
 * | **muteButton** | `boolean` | {@link ParticipantPanelItemMuteButtonDirective} |
 *
 * <p class="component-link-text">
 * <span class="italic">See all {@link ApiDirectiveModule API Directives}</span>
 * </p>
 *
 * </div>
 * <div>
 *
 * <h3>OpenVidu Angular Directives</h3>
 *
 * The ParticipantPanelItemComponent can be replaced with a custom component. It provides us the following {@link https://angular.io/guide/structural-directives Angular structural directives}
 * for doing this.
 *
 * |            **Directive**           |                 **Reference**                 |
 * |:----------------------------------:|:---------------------------------------------:|
 * |     ***ovParticipantPanelItem**     |     {@link ParticipantPanelItemDirective}     |
 *
 * </br>
 *
 * It is also providing us a way to **add additional buttons** to the default participant panel item.
 * It will recognise the following directive in a child element.
 *
 * |            **Directive**           |                 **Reference**                 |
 * |:----------------------------------:|:---------------------------------------------:|
 * | ***ovParticipantPanelItemElements** | {@link ParticipantPanelItemElementsDirective} |
 *
 * <p class="component-link-text">
 * 	<span class="italic">See all {@link OpenViduAngularDirectiveModule OpenVidu Angular Directives}</span>
 * </p>
 * </div>
 * </div>
 */
export class ParticipantPanelItemComponent {
    /**
     * @ignore
     */
    constructor(libService, participantService, cd) {
        this.libService = libService;
        this.participantService = participantService;
        this.cd = cd;
        /**
         * @ignore
         */
        this.showMuteButton = true;
    }
    /**
     * @ignore
     */
    set externalItemElements(externalItemElements) {
        // This directive will has value only when ITEM ELEMENTS component tagget with '*ovParticipantPanelItemElements' directive
        // is inside of the P PANEL ITEM component tagged with '*ovParticipantPanelItem' directive
        if (externalItemElements) {
            this.participantPanelItemElementsTemplate = externalItemElements.template;
        }
    }
    set participant(participant) {
        this._participant = participant;
    }
    ngOnInit() {
        this.subscribeToParticipantPanelItemDirectives();
    }
    ngOnDestroy() {
        if (this.muteButtonSub)
            this.muteButtonSub.unsubscribe();
    }
    /**
     * @ignore
     */
    toggleMuteForcibly() {
        this.participantService.setRemoteMutedForcibly(this._participant.id, !this._participant.isMutedForcibly);
    }
    subscribeToParticipantPanelItemDirectives() {
        this.muteButtonSub = this.libService.participantItemMuteButton.subscribe((value) => {
            this.showMuteButton = value;
            this.cd.markForCheck();
        });
    }
}
ParticipantPanelItemComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.8", ngImport: i0, type: ParticipantPanelItemComponent, deps: [{ token: i1.OpenViduAngularConfigService }, { token: i2.ParticipantService }, { token: i0.ChangeDetectorRef }], target: i0.ɵɵFactoryTarget.Component });
ParticipantPanelItemComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "14.2.8", type: ParticipantPanelItemComponent, selector: "ov-participant-panel-item", inputs: { participant: "participant" }, queries: [{ propertyName: "participantPanelItemElementsTemplate", first: true, predicate: ["participantPanelItemElements"], descendants: true, read: TemplateRef }, { propertyName: "externalItemElements", first: true, predicate: ParticipantPanelItemElementsDirective, descendants: true }], ngImport: i0, template: "<mat-list>\n\t<mat-list-item>\n\t\t<mat-icon matListAvatar class=\"participant-avatar\">person</mat-icon>\n\t\t<h3 matLine class=\"participant-nickname\">{{ _participant.nickname }}</h3>\n\t\t<p matLine class=\"participant-subtitle\">{{ _participant | streamTypesEnabled }}</p>\n\t\t<!-- <p matLine>\n\t\t\t<span class=\"participant-subtitle\"></span>\n\t\t</p> -->\n\n\t\t<div class=\"participant-action-buttons\">\n\t\t\t<button\n\t\t\t\tmat-icon-button\n\t\t\t\tid=\"mute-btn\"\n\t\t\t\t*ngIf=\"!_participant.local && showMuteButton\"\n\t\t\t\t[class.warn-btn]=\"_participant.isMutedForcibly\"\n\t\t\t\t(click)=\"toggleMuteForcibly()\"\n\t\t\t>\n\t\t\t\t<mat-icon *ngIf=\"!_participant.isMutedForcibly\">volume_up</mat-icon>\n\t\t\t\t<mat-icon *ngIf=\"_participant.isMutedForcibly\">volume_off</mat-icon>\n\t\t\t</button>\n\n\t\t\t<!-- External item elements  -->\n\t\t\t<ng-container *ngIf=\"participantPanelItemElementsTemplate\">\n\t\t\t\t<ng-container *ngTemplateOutlet=\"participantPanelItemElementsTemplate\"></ng-container>\n\t\t\t</ng-container>\n\t\t</div>\n\t</mat-list-item>\n</mat-list>\n", styles: [".participant-subtitle{font-style:italic;font-size:11px!important}.participant-nickname{font-weight:700!important}.participant-action-buttons{display:flex}::ng-deep .participant-action-buttons>*:not(#mute-btn){display:contents}::ng-deep .participant-action-buttons>*:not(#mute-btn)>*{margin:auto}mat-list-item{height:max-content!important;padding-bottom:10px!important}mat-list{padding:3px}.participant-avatar{display:contents}#mute-btn{border-radius:var(--ov-buttons-radius)}.warn-btn{color:var(--ov-warn-color)}\n"], dependencies: [{ kind: "directive", type: i3.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { kind: "directive", type: i3.NgTemplateOutlet, selector: "[ngTemplateOutlet]", inputs: ["ngTemplateOutletContext", "ngTemplateOutlet", "ngTemplateOutletInjector"] }, { kind: "component", type: i4.MatButton, selector: "button[mat-button], button[mat-raised-button], button[mat-icon-button],             button[mat-fab], button[mat-mini-fab], button[mat-stroked-button],             button[mat-flat-button]", inputs: ["disabled", "disableRipple", "color"], exportAs: ["matButton"] }, { kind: "component", type: i5.MatIcon, selector: "mat-icon", inputs: ["color", "inline", "svgIcon", "fontSet", "fontIcon"], exportAs: ["matIcon"] }, { kind: "directive", type: i6.MatLine, selector: "[mat-line], [matLine]" }, { kind: "component", type: i7.MatList, selector: "mat-list, mat-action-list", inputs: ["disableRipple", "disabled"], exportAs: ["matList"] }, { kind: "component", type: i7.MatListItem, selector: "mat-list-item, a[mat-list-item], button[mat-list-item]", inputs: ["disableRipple", "disabled"], exportAs: ["matListItem"] }, { kind: "directive", type: i7.MatListAvatarCssMatStyler, selector: "[mat-list-avatar], [matListAvatar]" }, { kind: "pipe", type: i8.StreamTypesEnabledPipe, name: "streamTypesEnabled" }], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.8", ngImport: i0, type: ParticipantPanelItemComponent, decorators: [{
            type: Component,
            args: [{ selector: 'ov-participant-panel-item', changeDetection: ChangeDetectionStrategy.OnPush, template: "<mat-list>\n\t<mat-list-item>\n\t\t<mat-icon matListAvatar class=\"participant-avatar\">person</mat-icon>\n\t\t<h3 matLine class=\"participant-nickname\">{{ _participant.nickname }}</h3>\n\t\t<p matLine class=\"participant-subtitle\">{{ _participant | streamTypesEnabled }}</p>\n\t\t<!-- <p matLine>\n\t\t\t<span class=\"participant-subtitle\"></span>\n\t\t</p> -->\n\n\t\t<div class=\"participant-action-buttons\">\n\t\t\t<button\n\t\t\t\tmat-icon-button\n\t\t\t\tid=\"mute-btn\"\n\t\t\t\t*ngIf=\"!_participant.local && showMuteButton\"\n\t\t\t\t[class.warn-btn]=\"_participant.isMutedForcibly\"\n\t\t\t\t(click)=\"toggleMuteForcibly()\"\n\t\t\t>\n\t\t\t\t<mat-icon *ngIf=\"!_participant.isMutedForcibly\">volume_up</mat-icon>\n\t\t\t\t<mat-icon *ngIf=\"_participant.isMutedForcibly\">volume_off</mat-icon>\n\t\t\t</button>\n\n\t\t\t<!-- External item elements  -->\n\t\t\t<ng-container *ngIf=\"participantPanelItemElementsTemplate\">\n\t\t\t\t<ng-container *ngTemplateOutlet=\"participantPanelItemElementsTemplate\"></ng-container>\n\t\t\t</ng-container>\n\t\t</div>\n\t</mat-list-item>\n</mat-list>\n", styles: [".participant-subtitle{font-style:italic;font-size:11px!important}.participant-nickname{font-weight:700!important}.participant-action-buttons{display:flex}::ng-deep .participant-action-buttons>*:not(#mute-btn){display:contents}::ng-deep .participant-action-buttons>*:not(#mute-btn)>*{margin:auto}mat-list-item{height:max-content!important;padding-bottom:10px!important}mat-list{padding:3px}.participant-avatar{display:contents}#mute-btn{border-radius:var(--ov-buttons-radius)}.warn-btn{color:var(--ov-warn-color)}\n"] }]
        }], ctorParameters: function () { return [{ type: i1.OpenViduAngularConfigService }, { type: i2.ParticipantService }, { type: i0.ChangeDetectorRef }]; }, propDecorators: { participantPanelItemElementsTemplate: [{
                type: ContentChild,
                args: ['participantPanelItemElements', { read: TemplateRef }]
            }], externalItemElements: [{
                type: ContentChild,
                args: [ParticipantPanelItemElementsDirective]
            }], participant: [{
                type: Input
            }] } });
