// openvidu-angular/esm2020/lib/services/participant/participant.service.mjs
import { Injectable } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import { ParticipantModel } from '../../models/participant.model';
import { VideoType } from '../../models/video-type.model';
import * as i0 from "@angular/core";
import * as i1 from "../config/openvidu-angular.config.service";
import * as i2 from "../logger/logger.service";
export class ParticipantService {
    /**
     * @internal
     */
    constructor(openviduAngularConfigSrv, loggerSrv) {
        this.openviduAngularConfigSrv = openviduAngularConfigSrv;
        this.loggerSrv = loggerSrv;
        this._localParticipant = new BehaviorSubject(null);
        this._remoteParticipants = new BehaviorSubject([]);
        this.remoteParticipants = [];
        this.log = this.loggerSrv.get('ParticipantService');
        this.localParticipantObs = this._localParticipant.asObservable();
        this.remoteParticipantsObs = this._remoteParticipants.asObservable();
    }
    /**
     * @internal
     */
    initLocalParticipant(props) {
        this.localParticipant = this.newParticipant(props);
        this.updateLocalParticipant();
    }
    getLocalParticipant() {
        return this.localParticipant;
    }
    /**
     * @internal
     */
    getMyCameraPublisher() {
        return this.localParticipant.getCameraConnection().streamManager;
    }
    /**
     * @internal
     */
    setMyCameraPublisher(publisher) {
        this.localParticipant.setCameraPublisher(publisher);
    }
    /**
     * @internal
     */
    setMyCameraConnectionId(connectionId) {
        this.localParticipant.setCameraConnectionId(connectionId);
    }
    /**
     * @internal
     */
    getMyScreenPublisher() {
        return this.localParticipant.getScreenConnection()?.streamManager;
    }
    /**
     * @internal
     */
    setMyScreenPublisher(publisher) {
        this.localParticipant.setScreenPublisher(publisher);
    }
    /**
     * @internal
     */
    setMyScreenConnectionId(connectionId) {
        this.localParticipant.setScreenConnectionId(connectionId);
    }
    /**
     * @internal
     */
    enableWebcamStream() {
        this.localParticipant.enableCamera();
        this.updateLocalParticipant();
    }
    /**
     * @internal
     */
    disableWebcamStream() {
        this.localParticipant.disableCamera();
        this.updateLocalParticipant();
    }
    /**
     * @internal
     */
    activeMyScreenShare(screenPublisher) {
        this.log.d('Enabling screen publisher');
        const steramModel = {
            type: VideoType.SCREEN,
            videoEnlarged: true,
            streamManager: screenPublisher,
            connected: true,
            connectionId: ''
        };
        this.resetRemoteStreamsToNormalSize();
        this.resetMyStreamsToNormalSize();
        this.localParticipant.addConnection(steramModel);
        this.updateLocalParticipant();
    }
    /**
     * @internal
     */
    disableScreenStream() {
        this.localParticipant.disableScreen();
        this.updateLocalParticipant();
    }
    /**
     * @internal
     */
    setMyNickname(nickname) {
        this.localParticipant.setNickname(nickname);
        this.updateLocalParticipant();
    }
    /**
     * @internal
     */
    getMyNickname() {
        return this.localParticipant.nickname;
    }
    getMyRole() {
        return this.localParticipant.getRole();
    }
    /**
     * @internal
     */
    toggleMyVideoEnlarged(connectionId) {
        this.localParticipant.toggleVideoEnlarged(connectionId);
    }
    /**
     * @internal
     */
    resetMyStreamsToNormalSize() {
        if (this.localParticipant.someHasVideoEnlarged()) {
            this.localParticipant.setAllVideoEnlarged(false);
            this.updateLocalParticipant();
        }
    }
    /**
     * @internal
     */
    clear() {
        this.disableScreenStream();
        this.remoteParticipants = [];
        this.updateRemoteParticipants();
        this.updateLocalParticipant();
    }
    /**
     * @internal
     */
    isMyCameraActive() {
        return this.localParticipant.isCameraActive();
    }
    isMyVideoActive() {
        return this.localParticipant.isCameraVideoActive();
    }
    isMyAudioActive() {
        return this.localParticipant?.hasAudioActive();
    }
    /**
     * @internal
     */
    isMyScreenActive() {
        return this.localParticipant.isScreenActive();
    }
    /**
     * @internal
     */
    isOnlyMyCameraActive() {
        return this.isMyCameraActive() && !this.isMyScreenActive();
    }
    /**
     * @internal
     */
    isOnlyMyScreenActive() {
        return this.isMyScreenActive() && !this.isMyCameraActive();
    }
    /**
     * @internal
     */
    haveICameraAndScreenActive() {
        return this.isMyCameraActive() && this.isMyScreenActive();
    }
    /**
     * @internal
     */
    hasScreenAudioActive() {
        return this.localParticipant.isScreenAudioActive();
    }
    /**
     * Force to update the local participant object and fire a new {@link localParticipantObs} Observable event.
     */
    updateLocalParticipant() {
        this._localParticipant.next(Object.assign(Object.create(this.localParticipant), this.localParticipant));
    }
    /**
     * REMOTE USERS
     */
    /**
     * @internal
     */
    addRemoteConnection(connectionId, data, subscriber) {
        const type = this.getTypeConnectionData(data);
        const streamModel = {
            type,
            videoEnlarged: type === VideoType.SCREEN,
            streamManager: subscriber,
            connected: true,
            connectionId
        };
        // Avoiding create a new participant if participantId param is not exist in connection data
        // participant Id is necessary for allowing to have multiple connection in one participant
        const participantId = this.getParticipantIdFromData(data) || connectionId;
        const participantAdded = this.getRemoteParticipantById(participantId);
        if (!!participantAdded) {
            this.log.d('Adding connection to existing participant: ', participantId);
            if (participantAdded.hasConnectionType(streamModel.type)) {
                this.log.d('Participant has publisher, updating it');
                participantAdded.setPublisher(streamModel.type, subscriber);
            }
            else {
                this.log.d('Participant has not publisher, adding it');
                if (streamModel.type === VideoType.SCREEN) {
                    this.resetRemoteStreamsToNormalSize();
                    this.resetMyStreamsToNormalSize();
                }
                participantAdded.addConnection(streamModel);
            }
        }
        else {
            this.log.w('Creating new participant with id: ', participantId);
            const props = {
                nickname: this.getNicknameFromConnectionData(data),
                local: false,
                id: participantId
            };
            const remoteParticipant = this.newParticipant(props, streamModel);
            this.remoteParticipants.push(remoteParticipant);
        }
        this.updateRemoteParticipants();
    }
    getRemoteParticipants() {
        return this.remoteParticipants;
    }
    /**
     * @internal
     */
    resetRemoteStreamsToNormalSize() {
        this.remoteParticipants.forEach((participant) => participant.setAllVideoEnlarged(false));
        this.updateRemoteParticipants();
    }
    /**
     * @internal
     */
    removeConnectionByConnectionId(connectionId) {
        this.log.w('Deleting connection: ', connectionId);
        let participant;
        if (this.localParticipant.hasConnectionId(connectionId)) {
            participant = this.localParticipant;
        }
        else {
            participant = this.getRemoteParticipantByConnectionId(connectionId);
        }
        if (participant) {
            const removeStream = participant.removeConnection(connectionId);
            //TODO: Timeout of X seconds?? Its possible sometimes the connections map was empty but must not be deleted
            if (participant.streams.size === 0) {
                // Remove participants without connections
                this.remoteParticipants = this.remoteParticipants.filter((p) => p !== participant);
            }
            if (removeStream.type === VideoType.SCREEN) {
                const remoteScreens = this.remoteParticipants.filter((p) => p.isScreenActive());
                if (remoteScreens.length > 0) {
                    // Enlarging the last screen connection active
                    const lastScreenActive = remoteScreens[remoteScreens.length - 1];
                    lastScreenActive.setScreenEnlarged(true);
                }
                else if (this.localParticipant.isScreenActive()) {
                    // Enlarging my screen if thereare not any remote screen active
                    this.localParticipant.setScreenEnlarged(true);
                }
            }
            this.updateRemoteParticipants();
        }
    }
    /**
     * @internal
     */
    getRemoteParticipantByConnectionId(connectionId) {
        return this.remoteParticipants.find((p) => p.hasConnectionId(connectionId));
    }
    getRemoteParticipantById(id) {
        return this.remoteParticipants.find((p) => p.id === id);
    }
    /**
     * @internal
     */
    someoneIsSharingScreen() {
        return this.remoteParticipants.some((p) => p.someHasVideoEnlarged());
    }
    /**
     * @internal
     */
    toggleRemoteVideoEnlarged(connectionId) {
        const participant = this.getRemoteParticipantByConnectionId(connectionId);
        participant?.toggleVideoEnlarged(connectionId);
    }
    /**
     * @internal
     */
    getNicknameFromConnectionData(data) {
        try {
            return JSON.parse(data).clientData;
        }
        catch (error) {
            return 'OpenVidu_User';
        }
    }
    /**
     * @internal
     */
    setRemoteNickname(connectionId, nickname) {
        const participant = this.getRemoteParticipantByConnectionId(connectionId);
        if (participant) {
            participant.setNickname(nickname);
            this.updateRemoteParticipants();
        }
    }
    /**
     * @internal
     */
    setRemoteMutedForcibly(id, value) {
        const participant = this.getRemoteParticipantById(id);
        if (participant) {
            participant.setMutedForcibly(value);
            this.updateRemoteParticipants();
        }
    }
    /**
     * Force to update the remote participants object and fire a new {@link remoteParticipantsObs} Observable event.
     */
    updateRemoteParticipants() {
        this._remoteParticipants.next([...this.remoteParticipants]);
    }
    /**
     * @internal
     * @param data
     * @returns Stream video type
     */
    getTypeConnectionData(data) {
        try {
            return JSON.parse(data).type;
        }
        catch (error) {
            return VideoType.CAMERA;
        }
    }
    getParticipantIdFromData(data) {
        try {
            return JSON.parse(data).participantId;
        }
        catch (error) {
            return '';
        }
    }
    newParticipant(props, streamModel) {
        if (this.openviduAngularConfigSrv.hasParticipantFactory()) {
            return this.openviduAngularConfigSrv.getParticipantFactory().apply(this, [props, streamModel]);
        }
        return new ParticipantModel(props, streamModel);
    }
}
ParticipantService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.8", ngImport: i0, type: ParticipantService, deps: [{ token: i1.OpenViduAngularConfigService }, { token: i2.LoggerService }], target: i0.ɵɵFactoryTarget.Injectable });
ParticipantService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "14.2.8", ngImport: i0, type: ParticipantService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.8", ngImport: i0, type: ParticipantService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: function () { return [{ type: i1.OpenViduAngularConfigService }, { type: i2.LoggerService }]; } });
