// openvidu-angular/esm2020/lib/services/caption/caption.service.mjs
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
import * as i0 from "@angular/core";
import * as i1 from "../storage/storage.service";
/**
 * @internal
 */
export class CaptionService {
    constructor(storageService) {
        this.storageService = storageService;
        this.langs = [
            { name: 'English', ISO: 'en-US' },
            { name: 'Español', ISO: 'es-ES' },
            { name: 'Deutsch', ISO: 'de-DE' },
            { name: 'Français', ISO: 'fr-FR' },
            { name: '中国', ISO: 'zh-CN' },
            { name: 'हिन्दी', ISO: 'hi-IN' },
            { name: 'Italiano', ISO: 'it-IT' },
            { name: 'やまと', ISO: 'jp-JP' },
            { name: 'Português', ISO: 'pt-PT' }
        ];
        this.captionLangSelected = { name: 'English', ISO: 'en-US' };
        this._captionLangObs = new Subject();
        this.captionsEnabled = false;
        const iso = this.storageService.getCaptionsLang();
        const lang = this.langs.find((lang) => lang.ISO === iso);
        if (iso && lang) {
            this.captionLangSelected = lang;
        }
        else {
            this.captionLangSelected = this.langs[0];
        }
        this.captionLangObs = this._captionLangObs.asObservable();
    }
    setCaptionsEnabled(value) {
        this.captionsEnabled = value;
    }
    areCaptionsEnabled() {
        return this.captionsEnabled;
    }
    setLanguage(lang) {
        const newLang = this.langs.find((l) => l.ISO === lang);
        if (!!newLang && newLang.ISO !== this.captionLangSelected.ISO) {
            this.captionLangSelected = newLang;
            this.storageService.setCaptionLang(lang);
            this._captionLangObs.next(this.captionLangSelected);
        }
    }
    getLangSelected() {
        return this.captionLangSelected;
    }
    getCaptionLanguages() {
        return this.langs;
    }
}
CaptionService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.8", ngImport: i0, type: CaptionService, deps: [{ token: i1.StorageService }], target: i0.ɵɵFactoryTarget.Injectable });
CaptionService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "14.2.8", ngImport: i0, type: CaptionService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.8", ngImport: i0, type: CaptionService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: function () { return [{ type: i1.StorageService }]; } });
