// openvidu-angular/esm2020/lib/directives/api/activities-panel.directive.mjs
import { Directive, Input } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "../../services/config/openvidu-angular.config.service";
/**
 * The **recordingActivity** directive allows show/hide the recording activity in {@link ActivitiesPanelComponent} activity panel component.
 *
 * Default: `true`
 *
 * It can be used in the parent element {@link VideoconferenceComponent} specifying the name of the `activitiesPanel` component:
 *
 * @example
 * <ov-videoconference [activitiesPanelRecordingActivity]="false"></ov-videoconference>
 *
 * \
 * And it also can be used in the {@link ActivitiesPanelComponent}.
 * @example
 * <ov-activities-panel [recordingActivity]="false"></ov-activities-panel>
 */
export class ActivitiesPanelRecordingActivityDirective {
    constructor(elementRef, libService) {
        this.elementRef = elementRef;
        this.libService = libService;
        this.recordingActivityValue = true;
    }
    set activitiesPanelRecordingActivity(value) {
        this.recordingActivityValue = value;
        this.update(this.recordingActivityValue);
    }
    set recordingList(value) {
        this.recordingActivityValue = value;
        this.update(this.recordingActivityValue);
    }
    ngAfterViewInit() {
        this.update(this.recordingActivityValue);
    }
    ngOnDestroy() {
        this.clear();
    }
    clear() {
        this.recordingActivityValue = true;
        this.update(true);
    }
    update(value) {
        if (this.libService.recordingActivity.getValue() !== value) {
            this.libService.recordingActivity.next(value);
        }
    }
}
ActivitiesPanelRecordingActivityDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.8", ngImport: i0, type: ActivitiesPanelRecordingActivityDirective, deps: [{ token: i0.ElementRef }, { token: i1.OpenViduAngularConfigService }], target: i0.ɵɵFactoryTarget.Directive });
ActivitiesPanelRecordingActivityDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "14.2.8", type: ActivitiesPanelRecordingActivityDirective, selector: "ov-videoconference[activitiesPanelRecordingActivity], ov-activities-panel[recordingActivity]", inputs: { activitiesPanelRecordingActivity: "activitiesPanelRecordingActivity", recordingList: "recordingList" }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.8", ngImport: i0, type: ActivitiesPanelRecordingActivityDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ov-videoconference[activitiesPanelRecordingActivity], ov-activities-panel[recordingActivity]'
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }, { type: i1.OpenViduAngularConfigService }]; }, propDecorators: { activitiesPanelRecordingActivity: [{
                type: Input
            }], recordingList: [{
                type: Input
            }] } });
