// openvidu-angular/esm2020/lib/directives/api/admin.directive.mjs
import { Directive, Input } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "../../services/config/openvidu-angular.config.service";
/**
 * The **recordingsList** directive allows show all recordings saved in your OpenVidu deployment in {@link AdminDashboardComponent}.
 *
 * Default: `[]`
 *
 * @example
 * <ov-admin-dashboard [recordingsList]="recordings"></ov-admin-dashboard>
 *
 */
export class AdminRecordingsListDirective {
    constructor(elementRef, libService) {
        this.elementRef = elementRef;
        this.libService = libService;
        this.recordingsValue = [];
    }
    set recordingsList(value) {
        this.recordingsValue = value;
        this.update(this.recordingsValue);
    }
    ngAfterViewInit() {
        this.update(this.recordingsValue);
    }
    ngOnDestroy() {
        this.clear();
    }
    clear() {
        this.recordingsValue = null;
        this.update(null);
    }
    update(value) {
        if (this.libService.adminRecordingsList.getValue() !== value) {
            this.libService.adminRecordingsList.next(value);
        }
    }
}
AdminRecordingsListDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.8", ngImport: i0, type: AdminRecordingsListDirective, deps: [{ token: i0.ElementRef }, { token: i1.OpenViduAngularConfigService }], target: i0.ɵɵFactoryTarget.Directive });
AdminRecordingsListDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "14.2.8", type: AdminRecordingsListDirective, selector: "ov-admin-dashboard[recordingsList]", inputs: { recordingsList: "recordingsList" }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.8", ngImport: i0, type: AdminRecordingsListDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ov-admin-dashboard[recordingsList]'
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }, { type: i1.OpenViduAngularConfigService }]; }, propDecorators: { recordingsList: [{
                type: Input
            }] } });
/**
 * The **error** directive allows show the authentication error in {@link AdminLoginComponent}.
 *
 * Default: `null`
 *
 * @example
 * <ov-admin-login [error]="error"></ov-admin-login>
 *
 */
export class AdminLoginDirective {
    constructor(elementRef, libService) {
        this.elementRef = elementRef;
        this.libService = libService;
        this.errorValue = null;
    }
    set error(value) {
        this.errorValue = value;
        this.update(this.errorValue);
    }
    ngAfterViewInit() {
        this.update(this.errorValue);
    }
    ngOnDestroy() {
        this.clear();
    }
    clear() {
        this.errorValue = null;
        this.update(null);
    }
    update(value) {
        if (this.libService.adminLoginError.getValue() !== value) {
            this.libService.adminLoginError.next(value);
        }
    }
}
AdminLoginDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.8", ngImport: i0, type: AdminLoginDirective, deps: [{ token: i0.ElementRef }, { token: i1.OpenViduAngularConfigService }], target: i0.ɵɵFactoryTarget.Directive });
AdminLoginDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "14.2.8", type: AdminLoginDirective, selector: "ov-admin-login[error]", inputs: { error: "error" }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.8", ngImport: i0, type: AdminLoginDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ov-admin-login[error]'
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }, { type: i1.OpenViduAngularConfigService }]; }, propDecorators: { error: [{
                type: Input
            }] } });
