// openvidu-angular/esm2020/lib/directives/api/participant-panel-item.directive.mjs
import { Directive, Input } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "../../services/config/openvidu-angular.config.service";
/**
 * The **muteButton** directive allows show/hide the muted button in participant panel item component.
 *
 * Default: `true`
 *
 * It can be used in the parent element {@link VideoconferenceComponent} specifying the name of the `participantPanelItem` component:
 *
 * @example
 * <ov-videoconference [participantPanelItemMuteButton]="false"></ov-videoconference>
 *
 * \
 * And it also can be used in the {@link ParticipantPanelItemComponent}.
 * @example
 * <ov-participant-panel-item [muteButton]="false"></ov-participant-panel-item>
 */
export class ParticipantPanelItemMuteButtonDirective {
    constructor(elementRef, libService) {
        this.elementRef = elementRef;
        this.libService = libService;
        this.muteValue = true;
    }
    set participantPanelItemMuteButton(value) {
        this.muteValue = value;
        this.update(this.muteValue);
    }
    set muteButton(value) {
        this.muteValue = value;
        this.update(this.muteValue);
    }
    ngAfterViewInit() {
        this.update(this.muteValue);
    }
    ngOnDestroy() {
        this.clear();
    }
    clear() {
        this.muteValue = true;
        this.update(true);
    }
    update(value) {
        if (this.libService.participantItemMuteButton.getValue() !== value) {
            this.libService.participantItemMuteButton.next(value);
        }
    }
}
ParticipantPanelItemMuteButtonDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.8", ngImport: i0, type: ParticipantPanelItemMuteButtonDirective, deps: [{ token: i0.ElementRef }, { token: i1.OpenViduAngularConfigService }], target: i0.ɵɵFactoryTarget.Directive });
ParticipantPanelItemMuteButtonDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "14.2.8", type: ParticipantPanelItemMuteButtonDirective, selector: "ov-videoconference[participantPanelItemMuteButton], ov-participant-panel-item[muteButton]", inputs: { participantPanelItemMuteButton: "participantPanelItemMuteButton", muteButton: "muteButton" }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.8", ngImport: i0, type: ParticipantPanelItemMuteButtonDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ov-videoconference[participantPanelItemMuteButton], ov-participant-panel-item[muteButton]'
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }, { type: i1.OpenViduAngularConfigService }]; }, propDecorators: { participantPanelItemMuteButton: [{
                type: Input
            }], muteButton: [{
                type: Input
            }] } });
