// openvidu-angular/esm2020/lib/directives/api/recording-activity.directive.mjs
import { Directive, Input } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "../../services/config/openvidu-angular.config.service";
/**
 * The **recordingsList** directive allows show the recordings available for the session in {@link RecordingActivityComponent}.
 *
 * Default: `[]`
 *
 * It can be used in the parent element {@link VideoconferenceComponent} specifying the name of the `recordingActivity` component:
 *
 * @example
 * <ov-videoconference [recordingActivityRecordingsList]="list"></ov-videoconference>
 *
 * \
 * And it also can be used in the {@link RecordingActivityComponent}.
 * @example
 * <ov-recording-activity [recordingsList]="list"></ov-recording-activity>
 */
export class RecordingActivityRecordingsListDirective {
    constructor(elementRef, libService) {
        this.elementRef = elementRef;
        this.libService = libService;
        this.recordingsValue = [];
    }
    set recordingActivityRecordingsList(value) {
        this.recordingsValue = value;
        this.update(this.recordingsValue);
    }
    set recordingsList(value) {
        this.recordingsValue = value;
        this.update(this.recordingsValue);
    }
    ngAfterViewInit() {
        this.update(this.recordingsValue);
    }
    ngOnDestroy() {
        this.clear();
    }
    clear() {
        this.recordingsValue = [];
        this.update([]);
    }
    update(value) {
        if (this.libService.recordingsList.getValue() !== value) {
            this.libService.recordingsList.next(value);
        }
    }
}
RecordingActivityRecordingsListDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.8", ngImport: i0, type: RecordingActivityRecordingsListDirective, deps: [{ token: i0.ElementRef }, { token: i1.OpenViduAngularConfigService }], target: i0.ɵɵFactoryTarget.Directive });
RecordingActivityRecordingsListDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "14.2.8", type: RecordingActivityRecordingsListDirective, selector: "ov-videoconference[recordingActivityRecordingsList], ov-recording-activity[recordingsList]", inputs: { recordingActivityRecordingsList: "recordingActivityRecordingsList", recordingsList: "recordingsList" }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.8", ngImport: i0, type: RecordingActivityRecordingsListDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ov-videoconference[recordingActivityRecordingsList], ov-recording-activity[recordingsList]'
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }, { type: i1.OpenViduAngularConfigService }]; }, propDecorators: { recordingActivityRecordingsList: [{
                type: Input
            }], recordingsList: [{
                type: Input
            }] } });
/**
 * The **recordingError** directive allows to show any possible error with the recording in the {@link RecordingActivityComponent}.
 *
 * Default: `[]`
 *
 * It can be used in the parent element {@link VideoconferenceComponent} specifying the name of the `recordingActivity` component:
 *
 * @example
 * <ov-videoconference [recordingActivityRecordingError]="error"></ov-videoconference>
 *
 * \
 * And it also can be used in the {@link RecordingActivityComponent}.
 * @example
 * <ov-recording-activity [recordingError]="error"></ov-recording-activity>
 */
export class RecordingActivityRecordingErrorDirective {
    constructor(elementRef, libService) {
        this.elementRef = elementRef;
        this.libService = libService;
        this.recordingErrorValue = null;
    }
    set recordingActivityRecordingError(value) {
        this.recordingErrorValue = value;
        this.update(this.recordingErrorValue);
    }
    set recordingError(value) {
        this.recordingErrorValue = value;
        this.update(this.recordingErrorValue);
    }
    ngAfterViewInit() {
        this.update(this.recordingErrorValue);
    }
    ngOnDestroy() {
        this.clear();
    }
    clear() {
        this.recordingErrorValue = null;
        this.update(null);
    }
    update(value) {
        if (this.libService.recordingError.getValue() !== value) {
            this.libService.recordingError.next(value);
        }
    }
}
RecordingActivityRecordingErrorDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.8", ngImport: i0, type: RecordingActivityRecordingErrorDirective, deps: [{ token: i0.ElementRef }, { token: i1.OpenViduAngularConfigService }], target: i0.ɵɵFactoryTarget.Directive });
RecordingActivityRecordingErrorDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "14.2.8", type: RecordingActivityRecordingErrorDirective, selector: "ov-videoconference[recordingActivityRecordingError], ov-recording-activity[recordingError]", inputs: { recordingActivityRecordingError: "recordingActivityRecordingError", recordingError: "recordingError" }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.8", ngImport: i0, type: RecordingActivityRecordingErrorDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ov-videoconference[recordingActivityRecordingError], ov-recording-activity[recordingError]'
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }, { type: i1.OpenViduAngularConfigService }]; }, propDecorators: { recordingActivityRecordingError: [{
                type: Input
            }], recordingError: [{
                type: Input
            }] } });
