// openvidu-angular/esm2020/lib/services/translate/translate.service.mjs
import { Injectable } from '@angular/core';
import * as cn from '../../lang/cn.json';
import * as de from '../../lang/de.json';
import * as en from '../../lang/en.json';
import * as es from '../../lang/es.json';
import * as fr from '../../lang/fr.json';
import * as hi from '../../lang/hi.json';
import * as it from '../../lang/it.json';
import * as ja from '../../lang/ja.json';
import * as nl from '../../lang/nl.json';
import * as pt from '../../lang/pt.json';
import * as i0 from "@angular/core";
import * as i1 from "../storage/storage.service";
/**
 * @internal
 */
export class TranslateService {
    constructor(storageService) {
        this.storageService = storageService;
        this.availableLanguages = { en, es, de, fr, cn, hi, it, ja, nl, pt };
        this.langTitles = [
            { name: 'English', ISO: 'en' },
            { name: 'Español', ISO: 'es' },
            { name: 'Deutsch', ISO: 'de' },
            { name: 'Français', ISO: 'fr' },
            { name: '中国', ISO: 'cn' },
            { name: 'हिन्दी', ISO: 'hi' },
            { name: 'Italiano', ISO: 'it' },
            { name: 'やまと', ISO: 'ja' },
            { name: 'Dutch', ISO: 'nl' },
            { name: 'Português', ISO: 'pt' }
        ];
        const iso = this.storageService.getLang() || 'en';
        this.langSelected = this.langTitles.find((lang) => lang.ISO === iso) || this.langTitles[0];
        this.currentLang = this.availableLanguages[this.langSelected.ISO];
    }
    setLanguage(lang) {
        if (this.langTitles.some(l => l.ISO === lang)) {
            this.currentLang = this.availableLanguages[lang];
            this.langSelected = this.langTitles.find((l) => l.ISO === lang);
        }
    }
    getLangSelected() {
        return this.langSelected;
    }
    getLanguagesInfo() {
        return this.langTitles;
    }
    translate(key) {
        let result = this.currentLang;
        key.split('.').forEach((prop) => {
            try {
                result = result[prop];
            }
            catch (error) {
                return '';
            }
        });
        return result;
    }
}
TranslateService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.8", ngImport: i0, type: TranslateService, deps: [{ token: i1.StorageService }], target: i0.ɵɵFactoryTarget.Injectable });
TranslateService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "14.2.8", ngImport: i0, type: TranslateService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.8", ngImport: i0, type: TranslateService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: function () { return [{ type: i1.StorageService }]; } });
