// openvidu-angular/esm2020/lib/services/storage/storage.service.mjs
import { Injectable } from '@angular/core';
import { Storage } from '../../models/storage.model';
import * as i0 from "@angular/core";
import * as i1 from "../logger/logger.service";
/**
 * @internal
 */
export class StorageService {
    constructor(loggerSrv) {
        this.loggerSrv = loggerSrv;
        this.storage = window.localStorage;
        this.log = this.loggerSrv.get('StorageService');
    }
    getNickname() {
        return this.get(Storage.USER_NICKNAME);
    }
    setNickname(name) {
        this.set(Storage.USER_NICKNAME, name);
    }
    getVideoDevice() {
        return this.get(Storage.VIDEO_DEVICE);
    }
    setVideoDevice(device) {
        this.set(Storage.VIDEO_DEVICE, device);
    }
    getAudioDevice() {
        return this.get(Storage.AUDIO_DEVICE);
    }
    setAudioDevice(device) {
        this.set(Storage.AUDIO_DEVICE, device);
    }
    isVideoMuted() {
        return this.get(Storage.VIDEO_MUTED) === 'true';
    }
    setVideoMuted(muted) {
        this.set(Storage.VIDEO_MUTED, `${muted}`);
    }
    isAudioMuted() {
        return this.get(Storage.AUDIO_MUTED) === 'true';
    }
    setAudioMuted(muted) {
        this.set(Storage.AUDIO_MUTED, `${muted}`);
    }
    setLang(lang) {
        this.set(Storage.LANG, lang);
    }
    getLang() {
        return this.get(Storage.LANG);
    }
    setCaptionLang(lang) {
        this.set(Storage.CAPTION_LANG, lang);
    }
    getCaptionsLang() {
        return this.get(Storage.CAPTION_LANG);
    }
    setBackground(id) {
        this.set(Storage.BACKGROUND, id);
    }
    getBackground() {
        return this.get(Storage.BACKGROUND);
    }
    removeBackground() {
        this.remove(Storage.BACKGROUND);
    }
    set(key, item) {
        const value = JSON.stringify({ item: item });
        // this.log.d('Storing on localStorage "' + key + '" with value "' + value + '"');
        this.storage.setItem(key, value);
    }
    get(key) {
        const str = this.storage.getItem(key);
        if (!!str) {
            return JSON.parse(str).item;
        }
        return null;
    }
    remove(key) {
        this.storage.removeItem(key);
    }
    clear() {
        this.log.d('Clearing localStorage');
        this.storage.clear();
    }
}
StorageService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.8", ngImport: i0, type: StorageService, deps: [{ token: i1.LoggerService }], target: i0.ɵɵFactoryTarget.Injectable });
StorageService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "14.2.8", ngImport: i0, type: StorageService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.8", ngImport: i0, type: StorageService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: function () { return [{ type: i1.LoggerService }]; } });
