// openvidu-angular/esm2020/lib/services/logger/logger.service.mjs
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "../config/openvidu-angular.config.service";
/**
 * @internal
 */
export class LoggerService {
    constructor(openviduAngularConfigSrv) {
        this.openviduAngularConfigSrv = openviduAngularConfigSrv;
        this.LOG_FNS = [];
        this.MSG_PREFIXES = [
            ['[', ']'],
            ['[', '] WARN: '],
            ['[', '] ERROR: ']
        ];
    }
    getLoggerFns(prefix) {
        this.log = window.console;
        this.LOG_FNS = [this.log.log, this.log.warn, this.log.error];
        const loggerFns = this.LOG_FNS.map((logTemplFn, i) => {
            return logTemplFn.bind(this.log, this.MSG_PREFIXES[i][0] + prefix + this.MSG_PREFIXES[i][1]);
        });
        return loggerFns;
    }
    get(prefix) {
        const prodMode = this.openviduAngularConfigSrv.isProduction();
        const loggerService = this;
        return {
            d: function (...args) {
                if (!prodMode) {
                    loggerService.getLoggerFns(prefix)[0].apply(this.log, arguments);
                }
            },
            w: function (...args) {
                loggerService.getLoggerFns(prefix)[1].apply(this.log, arguments);
            },
            e: function (...args) {
                loggerService.getLoggerFns(prefix)[2].apply(this.log, arguments);
            }
        };
    }
}
LoggerService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.8", ngImport: i0, type: LoggerService, deps: [{ token: i1.OpenViduAngularConfigService }], target: i0.ɵɵFactoryTarget.Injectable });
LoggerService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "14.2.8", ngImport: i0, type: LoggerService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.8", ngImport: i0, type: LoggerService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: function () { return [{ type: i1.OpenViduAngularConfigService }]; } });
